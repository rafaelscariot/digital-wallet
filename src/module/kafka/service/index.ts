export * from './kafka-consumer.service'
export * from './kafka-producer.service'