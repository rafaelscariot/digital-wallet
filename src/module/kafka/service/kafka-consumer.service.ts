import { Injectable, OnApplicationShutdown } from '@nestjs/common';
import {
  Consumer,
  ConsumerRunConfig,
  ConsumerSubscribeTopics,
  Kafka,
} from 'kafkajs';

@Injectable()
export class KafkaConsumerService implements OnApplicationShutdown {
  private readonly kafka = new Kafka({
    brokers: [process.env.KAFKA_URL],
  });

  private readonly consumers: Consumer[] = [];

  async consume(
    topic: ConsumerSubscribeTopics,
    config: ConsumerRunConfig,
  ): Promise<void> {
    const consumer = this.kafka.consumer({ groupId: 'digital-wallet' });
    await consumer.connect();
    await consumer.subscribe(topic);
    await consumer.run(config);
    this.consumers.push(consumer);
  }

  async onApplicationShutdown(): Promise<void> {
    for (const consumer of this.consumers) {
      await consumer.disconnect();
    }
  }
}
