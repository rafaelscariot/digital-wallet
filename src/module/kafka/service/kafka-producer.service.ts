import { OnModuleInit } from '@nestjs/common';
import { Injectable, OnApplicationShutdown } from '@nestjs/common';
import { Kafka, Producer, ProducerRecord } from 'kafkajs';

@Injectable()
export class KafkaProducerService
  implements OnApplicationShutdown, OnModuleInit
{
  private readonly kafka = new Kafka({
    brokers: ['localhost:9092'],
  });

  private readonly producer: Producer = this.kafka.producer();

  async onModuleInit() {
    await this.producer.connect();
  }

  async produce(record: ProducerRecord) {
    await this.producer.send(record);
  }

  async onApplicationShutdown() {
    await this.producer.disconnect();
  }
}
