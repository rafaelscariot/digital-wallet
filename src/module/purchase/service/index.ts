export * from './purchase.service';
export * from './cancel-purchase.service';
