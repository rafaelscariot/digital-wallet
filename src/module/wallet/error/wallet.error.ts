import { BadRequestException, NotFoundException } from '@nestjs/common';

export class WalletError {
  static NOT_FOUND(prop: string): NotFoundException {
    return new NotFoundException(`${prop.toUpperCase()}_NOT_FOUND`);
  }
  static INVALID_PAYLOAD(prop: string): BadRequestException {
    return new BadRequestException('INVALID_PAYLOAD');
  }
}
