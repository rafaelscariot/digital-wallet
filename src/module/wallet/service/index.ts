export * from './create-wallet.service';
export * from './get-wallet-balance-by-wallet-id.service';
export * from './get-wallet-statement-by-wallet-id.service';
export * from './deposit.service';
