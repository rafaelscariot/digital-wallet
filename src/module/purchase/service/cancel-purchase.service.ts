import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '@db/service';
import { KafkaProducerService } from '@kafka/service';
import { TopicEnum } from '@shared/enum';

@Injectable()
export class CancelPurchaseService {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly kafkaProducerService: KafkaProducerService,
  ) {}

  private readonly logger = new Logger(CancelPurchaseService.name);

  async perform(payload: string): Promise<void> {
    try {
      const { purchaseId } = await this.parsePayload(payload);

      this.logger.log(`Cancelling the purchase ${purchaseId}`);

      const purchase = await this.prismaService.purchase.findUnique({
        where: { id: purchaseId },
      });

      if (!purchase) {
        throw Error(`Purchase ${purchaseId} not found`);
      }

      if (purchase.canceled) {
        throw Error(`Purchase ${purchaseId} already canceled`);
      }

      const purchaseWallet = await this.prismaService.wallet.findUnique({
        where: { id: purchase.walletId },
      });

      const updatePurchase = this.prismaService.purchase.update({
        where: { id: purchase.id },
        data: { canceled: true },
      });

      const updateWallet = this.prismaService.wallet.update({
        data: {
          amount: Number(purchaseWallet.amount) + Number(purchase.amount),
        },
        where: { id: purchaseWallet.id },
      });

      const createWalletStatement = this.prismaService.walletStatement.create({
        data: {
          walletId: purchaseWallet.id,
          reversal: purchase.amount,
        },
      });

      await this.prismaService.$transaction([
        updatePurchase,
        updateWallet,
        createWalletStatement,
      ]);
    } catch (error) {
      this.logger.error(error);

      await this.kafkaProducerService.produce({
        topic: 'error',
        messages: [
          {
            value: {
              topic: TopicEnum.CANCELLATION,
              payload,
              error: error,
            }.toString(),
          },
        ],
      });
    }
  }

  private async parsePayload(payload: string): Promise<{
    purchaseId: number;
  }> {
    const { purchaseId } = JSON.parse(payload);

    if (!purchaseId) throw Error('Payload missing attributes');

    return { purchaseId: Number(purchaseId) };
  }
}
