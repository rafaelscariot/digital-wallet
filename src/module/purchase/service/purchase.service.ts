import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '@db/service';
import { KafkaProducerService } from '@kafka/service';
import { TopicEnum } from '@shared/enum';

@Injectable()
export class PurchaseService {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly kafkaProducerService: KafkaProducerService,
  ) {}

  private readonly logger = new Logger(PurchaseService.name);

  async perform(payload: string): Promise<void> {
    try {
      const { walletId, amount } = await this.parsePayload(payload);

      this.logger.log(`Purchase of ${amount.toFixed(2)} in the wallet ${walletId}`);

      const wallet = await this.prismaService.wallet.findUnique({
        where: { id: walletId },
      });

      if (!wallet) {
        throw Error(`Wallet ${walletId} not found`);
      }

      if (amount > Number(wallet.amount)) {
        throw Error('Invalid amount');
      }

      const updateWallet = this.prismaService.wallet.update({
        data: { amount: Number(wallet.amount) - amount },
        where: { id: walletId },
      });

      const createPurchase = this.prismaService.purchase.create({
        data: { walletId, amount },
      });

      const createWalletStatement = this.prismaService.walletStatement.create({
        data: { walletId, withdrawal: amount },
      });

      await this.prismaService.$transaction([
        updateWallet,
        createPurchase,
        createWalletStatement,
      ]);
    } catch (error) {
      this.logger.error(error);

      await this.kafkaProducerService.produce({
        topic: 'error',
        messages: [
          {
            value: {
              topic: TopicEnum.PURCHASE,
              payload,
              error: error,
            }.toString(),
          },
        ],
      });
    }
  }

  private async parsePayload(payload: string): Promise<{
    walletId: number;
    amount: number;
  }> {
    const { walletId, amount } = JSON.parse(payload);

    if (!walletId || !amount) throw Error('Payload missing attributes');

    return { walletId: Number(walletId), amount: Number(amount) };
  }
}
