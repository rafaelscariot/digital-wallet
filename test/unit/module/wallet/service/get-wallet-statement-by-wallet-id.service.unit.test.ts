import { suite, test } from '@testdeck/jest';
import { GetWalletStatementByWalletIdService } from '@wallet/service';
import { PrismaService } from '@db/service';
import { WalletError } from '@wallet/error';

@suite('[Wallet Module] Get Wallet Statement By Wallet Id Service Unit Test')
class GetWalletStatementByWalletIdServiceUnitTest {
  private getWalletStatementByWalletIdService: GetWalletStatementByWalletIdService;
  private prismaServiceMock: PrismaService;

  async before() {
    this.prismaServiceMock = {
      wallet: {
        findUnique: ({ where: { id } }) => {
          return { walletId: 1, amount: 10 };
        },
      },
      walletStatement: {
        findMany: ({ where: { walletId }, orderBy: { createdAt } }) => {
          return {
            amount: 10,
            movements: [
              {
                createdAt: new Date('Jul 12 2011'),
                walletId,
                deposit: 10,
                withdrawal: null,
                reversal: null,
              },
            ],
          };
        },
      },
    } as unknown as PrismaService;

    this.getWalletStatementByWalletIdService =
      new GetWalletStatementByWalletIdService(this.prismaServiceMock);
  }

  @test
  async 'Given a valid wallet id, should return the respective statements'() {
    const statements = await this.getWalletStatementByWalletIdService.perform(
      1,
    );

    expect(statements).toMatchSnapshot();
  }

  @test
  async 'Given an valid wallet id, should return an error'() {
    this.prismaServiceMock.wallet.findUnique = ({ where: { id } }) => {
      return null;
    };
    const statements = this.getWalletStatementByWalletIdService.perform(1);
    await expect(statements).rejects.toThrow(
      WalletError.NOT_FOUND(`wallet_${1}`),
    );
  }
}
