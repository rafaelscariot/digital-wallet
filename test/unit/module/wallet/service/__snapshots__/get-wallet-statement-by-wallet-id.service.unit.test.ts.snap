// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Wallet Module] Get Wallet Statement By Wallet Id Service Unit Test Given a valid wallet id, should return the respective statements 1`] = `
{
  "amount": 10,
  "movements": {
    "amount": 10,
    "movements": [
      {
        "createdAt": 2011-07-12T03:00:00.000Z,
        "deposit": 10,
        "reversal": null,
        "walletId": 1,
        "withdrawal": null,
      },
    ],
  },
}
`;
